chain-extensions/pallet-assets/src/tests.rs

```diff
@@ -0,0 +1,478 @@
+        });
+}
+
+#[test]
+fn burn_works() {
+    ExtBuilder::default()
+        .existential_deposit(50)
+        .build()
+        .execute_with(|| {
+            let addr = instantiate();
+
+            // Arrange - create asset and give contract mint permission (Issuer role) and burn permission (Admin role)
+            assert_ok!(Assets::create(
+                RuntimeOrigin::signed(ALICE),
+                ASSET_ID,
+                ALICE,
+                1
+            ));
+            assert_ok!(Assets::set_team(
+                RuntimeOrigin::signed(ALICE),
+                ASSET_ID,
+                addr.clone(),
+                addr.clone(),
+                ALICE
+            ));
+
+            // Act - Mint 1000 assets to Alice
+            assert_ok!(mint(addr.clone(), ASSET_ID, ALICE, 1000));
+            assert_eq!(Assets::balance(ASSET_ID, ALICE), 1000);
+
+            // Act - Burn 1000 of Alice tokens
+            assert_ok!(burn(addr.clone(), ASSET_ID, ALICE, 1000));
+
+            // Assert - Balance of Alice is then 0
+            assert_eq!(Assets::balance(ASSET_ID, ALICE), 0);
+        });
+}
+
+#[test]
+fn transfer_works() {
+    ExtBuilder::default()
+        .existential_deposit(50)
+        .build()
+        .execute_with(|| {
+            let addr = instantiate();
+
+            // Assert - Create, mint and transfer 1000 to contract
+            assert_ok!(Assets::create(RuntimeOrigin::signed(BOB), ASSET_ID, BOB, 1));
+            assert_ok!(Assets::mint(
+                RuntimeOrigin::signed(BOB),
+                ASSET_ID,
+                BOB,
+                1000
+            ));
+            assert_ok!(Assets::transfer(
+                RuntimeOrigin::signed(BOB),
+                ASSET_ID,
+                addr.clone(),
+                1000
+            ));
+
+            // Act - Transfer 1000 from contract to Alice
+            assert_ok!(transfer(addr.clone(), ASSET_ID, ALICE, 1000));
+
+            // Assert - Alice balance is 1000 and contract is zero
+            assert_eq!(Assets::balance(ASSET_ID, ALICE), 1000);
+            assert_eq!(Assets::balance(ASSET_ID, addr.clone()), 0);
+        });
+}
+
+#[test]
+fn balance_of_and_total_supply() {
+    ExtBuilder::default()
+        .existential_deposit(50)
+        .build()
+        .execute_with(|| {
+            let addr = instantiate();
+
+            // Arrange - create & mint 1000 to Alice
+            assert_ok!(Assets::create(
+                RuntimeOrigin::signed(ALICE),
+                ASSET_ID,
+                ALICE,
+                1
+            ));
+            assert_ok!(Assets::mint(
+                RuntimeOrigin::signed(ALICE),
+                ASSET_ID,
+                ALICE,
+                1000
+            ));
+
+            // Assert - Balance and total supply is 1000
+            assert_eq!(
+                balance_of(addr.clone(), ASSET_ID, ALICE).data[1..],
+                1000u128.encode()
```

Comment:
I guess you are truncating the 1st byte for the removing the result encoded byte?
IMO more readable approach would be decoding into result and match.
```rust
assert_eq!(
      Result::<u128,_>::decode(balance_of(addr.clone(), ASSET_ID, ALICE).data).unwrap(),
      000u128
);
```

This is informational you ignore it if you want

Reply:
It's not a result it's data field of ExecReturnValue. But the first byte is an empty byte so I just read buffer from second byte.
Haven't find why this first byte is an empty byte but seems in XVM [tests](https://github.com/AstarNetwork/Astar/blob/21acad96ef65648919df302a6c07a5bfaefadea0/tests/integration/src/xvm.rs#L809C14-L810C41) it's done the same